src/main/kotlin/com/exactpro/th2/codec/Application.kt

```diff
@@ -83,69 +84,85 @@ class Application(commonFactory: CommonFactory): AutoCloseable {
 
     init {
         val grpcRouter: GrpcRouter = commonFactory.grpcRouter
-        val grpcDecoder = createProtoDecoder("grpc-decoder", "", "", true).apply { codecs += this }
-        val grpcEncoder = createProtoEncoder("grpc-encoder", "", "", true).apply { codecs += this }
-        val grpcService = GrpcCodecService(grpcRouter, grpcDecoder::handleMessage, grpcEncoder::handleMessage, configuration.isFirstCodecInPipeline, eventProcessor)
+        val grpcDecoder = ProtoSyncDecoder(
+            eventRouter,
+            ProtoDecodeProcessor(context.codec, context.protocols, true, configuration.enableVerticalScaling, eventProcessor),
+            rootEventId
+        )
+        val grpcEncoder = ProtoSyncEncoder(
+            eventRouter,
+            ProtoEncodeProcessor(context.codec, context.protocols, true, configuration.enableVerticalScaling, eventProcessor),
+            rootEventId
+        )
+        val grpcService = GrpcCodecService(grpcRouter, grpcDecoder::handleBatch, grpcEncoder::handleBatch, configuration.isFirstCodecInPipeline, eventProcessor)
         grpcServer = grpcRouter.startServer(grpcService)
         grpcServer.start()
         K_LOGGER.info { "codec started" }
     }
 
-    private fun createProtoEncoder(
+    private fun createMqProtoEncoder(
         codecName: String,
         sourceAttributes: String,
         targetAttributes: String,
         useParentEventId: Boolean
-    ): AbstractProtoSyncCodec = ProtoSyncEncoder(
-            protoRouter,
+    ): AutoCloseable = MqListener(
+        protoRouter,
+        ProtoSyncEncoder(
             eventRouter,
             ProtoEncodeProcessor(context.codec, context.protocols, useParentEventId, configuration.enableVerticalScaling, eventProcessor),
             rootEventId
-        ).apply {
-            start(sourceAttributes, targetAttributes)
-        }.also { K_LOGGER.info { "Proto '$codecName' started" } }
+        )::handleBatch,
+        sourceAttributes,
+        targetAttributes
+    ).also { K_LOGGER.info { "Proto '$codecName' started" } }
 
-    private fun createProtoDecoder(
+    private fun createMqProtoDecoder(
         codecName: String,
         sourceAttributes: String,
         targetAttributes: String,
         useParentEventId: Boolean
-    ): AbstractProtoSyncCodec = ProtoSyncDecoder(
-            protoRouter,
+    ): AutoCloseable = MqListener(
+        protoRouter,
+        ProtoSyncDecoder(
             eventRouter,
             ProtoDecodeProcessor(context.codec, context.protocols, useParentEventId, configuration.enableVerticalScaling, eventProcessor),
             rootEventId
-        ).apply {
-            start(sourceAttributes, targetAttributes)
-        }.also { K_LOGGER.info { "Proto '$codecName' started" } }
+        )::handleBatch,
```

Comment:
I think the creation of `ProtoSyncDecoder/Encoder` can be extracted into a separate method and reused when we create a decoder/encoder for gRPC